Code/Framework/AzToolsFramework/Tests/Viewport/ViewportEditorModeTests.cpp

```diff
@@ -0,0 +1,492 @@
+        }
+
+        ~ViewportEditorModeNotificationsBusHandler()
+        {
+            AzToolsFramework::ViewportEditorModeNotificationsBus::Handler::BusDisconnect();
+        }
+
+        ViewportId GetViewportSubscription() const
+        {
+            return m_viewportSubscription;
+        }
+
+        const EditModeTracker& GetEditorModes() const
+        {
+            return m_editorModes;
+        }
+
+        void OnEditorModeEnter([[maybe_unused]]const ViewportEditorModesInterface& editorModeState, ViewportEditorMode mode) override
+        {
+            m_editorModes[mode].m_onEnter = true;
+        }
+
+        virtual void OnEditorModeExit([[maybe_unused]] const ViewportEditorModesInterface& editorModeState, ViewportEditorMode mode) override
+        {
+            m_editorModes[mode].m_onLeave = true;
+        }
+
+     private:
+         ViewportId m_viewportSubscription;
+         EditModeTracker m_editorModes;
+
+    };
+
+    // Fixture for testing viewport editor mode notifications publishing
+    class ViewportEditorModePublisherTestFixture
+        : public ViewportEditorModeTrackerTestFixture
+    {
+    public:
+
+        void SetUpEditorFixtureImpl() override
+        {
+            for (auto mode = 0; mode < ViewportEditorModes::NumEditorModes; mode++)
+            {
+                m_editorModeHandlers[mode] = AZStd::make_unique<ViewportEditorModeNotificationsBusHandler>(mode);
+            }
+        }
+
+        void TearDownEditorFixtureImpl() override
+        {
+            for (auto mode = 0; mode < ViewportEditorModes::NumEditorModes; mode++)
+            {
+                m_editorModeHandlers[mode].reset();
+            }
+        }
+
+        AZStd::array<AZStd::unique_ptr<ViewportEditorModeNotificationsBusHandler>, ViewportEditorModes::NumEditorModes> m_editorModeHandlers;
+    };
+
+    TEST_F(ViewportEditorModesTestsFixture, NumberOfEditorModesIsEqualTo4)
+    {
+        EXPECT_EQ(ViewportEditorModes::NumEditorModes, 4);
+    }
+
+    TEST_F(ViewportEditorModesTestsFixture, InitialEditorModeStateHasAllInactiveModes)
+    {
+        for (auto mode = 0; mode < ViewportEditorModes::NumEditorModes; mode++)
+        {
+            EXPECT_FALSE(m_editorModes.IsModeActive(static_cast<ViewportEditorMode>(mode)));
+        }
+    }
+
+    TEST_P(ViewportEditorModesTestsFixtureWithParams, SettingModeActiveActivatesOnlyThatMode)
+    {
+        m_editorModes.SetModeActive(m_selectedEditorMode);
+
+        for (auto mode = 0; mode < ViewportEditorModes::NumEditorModes; mode++)
+        {
+            const auto editorMode = static_cast<ViewportEditorMode>(mode);
+            if (editorMode == m_selectedEditorMode)
+            {
+                EXPECT_TRUE(m_editorModes.IsModeActive(static_cast<ViewportEditorMode>(editorMode)));
+            }
+            else
+            {
+                EXPECT_FALSE(m_editorModes.IsModeActive(static_cast<ViewportEditorMode>(editorMode)));
+            }
+        }
```

Comment:
suggestion: It might be possible to rewrite this with a matcher to be more succinct (e.g. ElementsAre, ElementsAreArray etc...). I've found avoiding loops in tests is a good thing to strive for so only a single call to an expectation is required (if this is too difficult/time consuming to change now this is fine, but might be worth having a look into to see if it can be rewritten and simplified.

Reply:
I'll look into this 👍

Reply:
Nice once thanks, this page is a treasure trove - http://google.github.io/googletest/reference/matchers.html (also there's some info on a fairly recent ly-goodreads article on the topic too)